### PR TITLE
CLOUDSTACK-9203 Implement security group move on updateVM API call

### PR DESCRIPTION
**Pull request by DaanHoogland**
cherry-picked from a exoscale internal fix

Conflicts:
	api/src/org/apache/cloudstack/api/command/user/vm/UpdateVMCmd.java
	server/src/com/cloud/vm/UserVmManager.java
	server/src/com/cloud/vm/UserVmManagerImpl.java

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@borisroman @wido as of now this is just a place holder as some extra requirements are put in by @resmo and no testing is done yet. I will update with at least the extra parameter and hopefully some test data. @resmo can you help with the testing? Thanks @llambiel and @pyr for the contribution.

**Comment by DaanHoogland** [comment]
ping @ustcweizhou

**Comment by DaanHoogland** [comment]
done some testing:

screen shot 2015-12-31 at 22 55 55
screen shot 2015-12-31 at 22 56 20
with cloudmonkey:
```
🐵 > update virtualmachine id=39f058ad-3021-4753-9c96-1aa922a74632 securitygroupnames=exceptional
Error 431: Virtual machine must be stopped prior to update security groups

🐵 > stop virtualmachine id=39f058ad-3021-4753-9c96-1aa922a74632
...
🐵 > update virtualmachine id=39f058ad-3021-4753-9c96-1aa922a74632 securitygroupnames=exceptional
...
```
resulting in
screen shot 2015-12-31 at 22 58 54

**Comment by DaanHoogland** [comment]
pinging every badoy again: @borisroman @wido @resmo @llambiel @pyr @bhaisaab @wilderrodrigues

**Comment by resmo** [comment]
LGTM

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -2288,6 +2290,7 @@ public UserVm updateVirtualMachine(UpdateVMCmd cmd) throws ResourceUnavailableEx
         String hostName = cmd.getHostName();
         Map<String,String> details = cmd.getDetails();
         Account caller = CallContext.current().getCallingAccount();
+        List<Long> securityGroupIdList = getSecurityGroupIdList(cmd);
```
Comment (wido) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: Shouldn't there be a check for NULL here?

**Comment by wido** [comment]
One comment I have on the code. Otherwise it looks good.

File: server/src/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (DaanHoogland) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: The updateVirtualMachine below can handle the null value (see line 2463)

**Comment by wido** [comment]
Checked all the commits, re-read the code and LGTM. Based on the tests already performed.

**Comment by DaanHoogland** [comment]
@borisroman @wido @resmo @llambiel @pyr @bhaisaab @wilderrodrigues I can run the tests again but I am the contributor. Can someone else, please?

**Comment by ustcweizhou** [comment]
this does not work if there are multiple shared network in advanced zone with sg.
I will fix it.

**Comment by DaanHoogland** [comment]
@wido @remibergsma Can we merge?

**Comment by remibergsma** [comment]
@DaanHoogland Did I miss the integration tests?

**Comment by DaanHoogland** [comment]
@remibergsma the api was tested in sg. We can run a full regression suit if you feel it is needed. I don't feel we need to test anything but the call with updated security groups.

**Comment by DaanHoogland** [comment]
btw it also is based on exoscale production code

**Comment by ustcweizhou** [comment]
LGTM, based on manual functional testing in advanced zone with security groups.

**Comment by wido** [comment]
Can we rebase against master (4.9)? Seems that this one is ready to merge

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 137
 Hypervisor xenserver
 NetworkType Advanced
 Passed=101
 Failed=1
 Skipped=4

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* integration.smoke.test_privategw_acl.TestPrivateGwACL

 * test_01_vpc_privategw_acl Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_deploy_vgpu_enabled_vm
test_06_copy_template
test_06_copy_iso

**Passed test suits:**
integration.smoke.test_deploy_vm_with_userdata.TestDeployVmWithUserData
integration.smoke.test_affinity_groups_projects.TestDeployVmWithAffinityGroup
integration.smoke.test_portable_publicip.TestPortablePublicIPAcquire
integration.smoke.test_over_provisioning.TestUpdateOverProvision
integration.smoke.test_global_settings.TestUpdateConfigWithScope
integration.smoke.test_scale_vm.TestScaleVm
integration.smoke.test_service_offerings.TestCreateServiceOffering
integration.smoke.test_loadbalance.TestLoadBalance
integration.smoke.test_routers.TestRouterServices
integration.smoke.test_reset_vm_on_reboot.TestResetVmOnReboot
integration.smoke.test_snapshots.TestSnapshotRootDisk
integration.smoke.test_deploy_vms_with_varied_deploymentplanners.TestDeployVmWithVariedPlanners
integration.smoke.test_network.TestDeleteAccount
integration.smoke.test_non_contigiousvlan.TestUpdatePhysicalNetwork
integration.smoke.test_deploy_vm_iso.TestDeployVMFromISO
integration.smoke.test_public_ip_range.TestDedicatePublicIPRange
integration.smoke.test_multipleips_per_nic.TestDeployVM
integration.smoke.test_regions.TestRegions
integration.smoke.test_affinity_groups.TestDeployVmWithAffinityGroup
integration.smoke.test_network_acl.TestNetworkACL
integration.smoke.test_pvlan.TestPVLAN
integration.smoke.test_volumes.TestCreateVolume
integration.smoke.test_ssvm.TestSSVMs
integration.smoke.test_nic.TestNic
integration.smoke.test_deploy_vm_root_resize.TestDeployVM
integration.smoke.test_resource_detail.TestResourceDetail
integration.smoke.test_secondary_storage.TestSecStorageServices
integration.smoke.test_disk_offerings.TestCreateDiskOffering

**Comment by NuxRo** [comment]
Daan,

How did you test?
I'm trying to apply the patch to either 4.8.0 or master and on neither it does so cleanly:

Hunk #7 FAILED at 467.
1 out of 7 hunks FAILED -- saving rejects to file api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java.rej

**Comment by DaanHoogland** [comment]
@nux, it probably doesn't apply anymore. This PR is 4 months old :(

**Comment by swill** [comment]
@NuxRo asked that I review this.  @DaanHoogland I see you opened this, would you mind rebasing.  I suspect it is still valid considering the fact that not a huge amount of code has been merged since you opened this.

**Comment by DaanHoogland** [comment]
new integration run pending

**Comment by DaanHoogland** [comment]
### CI bubble run


**Commit Reference: de53cd1799122f6448faef950cf9afd5caeaf412**


initially two test failed. I reran the 'unusual' failure and it succeeded.

**Comment by DaanHoogland** [comment]
@ustcweizhou Can you review, the rebasing led to more commits then expected. The result looks fine but I am not sure.

**Comment by DaanHoogland** [comment]
@remibergsma @swill this one test always failing in our environments doesn't sit good with me. It feels like we are accepting a crippled system to be carried. I'll spend some time there.

**Comment by swill** [comment]
@DaanHoogland for the test that often fails in my environment, check @remibergsma's comments here: https://github.com/apache/cloudstack/pull/1449#issuecomment-204764789

**Comment by DaanHoogland** [comment]
@swill that is definitely not the same issue as I'm having. it is in the same test however

**Comment by DaanHoogland** [comment]
@swill sorry it is not the same test that fails in my env: 


"Ping to outside world from VM should be successful" in test_02_redundant_VPC_default_routes is mine

**Comment by NuxRo** [comment]
Applied successfully to 4.8 and tested, it works fine as far as I could tell, but the SG change operation goes through even if the VM is online, so "CLOUDSTACK-9203 Prevent security group update while instance is running" is broken?

**Comment by rhtyd** [comment]
@DaanHoogland rebase against master, squash changes into a single commit, thanks

Please advise how to test this

**Comment by NuxRo** [comment]
rhtyd,

You can test via cloudmonkey, e.g.
update virtualmachine id=937daff4-7ce8-454c-bc8b-6f96b72017f5 securitygroupids=db6b797f-5262-4be2-a3b1-2a86c7c4f77b,98c9f1da-ce7a-11e5-a710-f2a3ece198a5

**Comment by DaanHoogland** [comment]
@rhtyd rebased, as these are several different commits by different people from different companies, I will not squash it.

**Comment by rhtyd** [comment]
@DaanHoogland thanks, how about you have consecutive changes by the same author squashed to one commit; while keep it separate per author in general

**Comment by rhtyd** [comment]
@DaanHoogland can you do a rebase and push -f, Travis failed for `test_01_scale_vm` test

In general, LGTM (just code review)

**Comment by NuxRo** [comment]
Any idea why Travis fails that 1 job?

**Comment by rhtyd** [comment]
Looks like a cleanup/env issue, can you push -f again @DaanHoogland 

In Travis, scroll down at the end -- see the table which test failed (exception/failed etc) and at the end click on the second last knob/triangle to see marvin output and search for failed test:

```
2016-05-03 17:31:02,554 - CRITICAL - EXCEPTION: test_01_scale_vm: ['Traceback (most recent call last):\n', '  File "/opt/python/2.7.10/lib/python2.7/unittest/case.py", line 331, in run\n    testMethod()\n', '  File "/home/travis/build/apache/cloudstack/test/integration/smoke/test_scale_vm.py", line 151, in test_01_scale_vm\n    isdynamicallyscalable=\'true\')\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/lib/base.py", line 631, in update\n    return(apiclient.updateVirtualMachine(cmd))\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/cloudstackAPI/cloudstackAPIClient.py", line 740, in updateVirtualMachine\n    response = self.connection.marvinRequest(command, response_type=response, method=method)\n', '  File "/home/travis/.local/lib/python2.7/site-packages/marvin/cloudstackConnection.py", line 379, in marvinRequest\n    raise e\n', 'CloudstackAPIException: Execute cmd: updatevirtualmachine failed, due to: errorCode: 431, errorText:Unable to find network with trafficType Guest and guestType Shared in zone 1\n']
```

**Comment by NuxRo** [comment]
Ah, indeed. It should be running the test against a Security Group zone, not Adv.

**Comment by DaanHoogland** [comment]
@rhtyd @NuxRo sound like a regession. I will look at it.

**Comment by NuxRo** [comment]
Sorry to be a pain, but any updates on this?

**Comment by DaanHoogland** [comment]
@nuxro I have rebased but did not get to testing. I don't trust the scalevm problem. It might be in the test, in the test environment or in the actual code. Looking at it is in (the back of) my queue.

**Comment by NuxRo** [comment]
Thanks Daan

**Comment by DaanHoogland** [comment]
@ don't thank me, I just started to check but my infra is based on KVM and it doesn't support scalevm, so I cannot investigate further

**Comment by NuxRo** [comment]
Same here.

Is this XenServer specific, can anyone with an XS test bed help?

**Comment by swill** [comment]
@DaanHoogland can you force push as we should have fixes for Jenkins and Travis in place now.  If we can get some final verification on this one then we can get into 4.9.

**Comment by DaanHoogland** [comment]
I'm an ass but no. I can close and reopen however. should do the trick

**Comment by swill** [comment]
This one errored again in both Jenkins and Travis.  Not sure if that signifies an actual problem or if things are just acting up.  I fix to some of the issues in Jenkins and Travis have been pushed since this was last run.  Can we kick this one again to see if it comes back clean now.  If not, we may need to investigate.  Thx...

**Comment by NuxRo** [comment]
It still doesn't look. I've tested this in a real setup and it seems to do what it's supposed to.

Anything I can help with to get past these errors?

**Comment by DaanHoogland** [comment]
@NuxRo I am running the integration suite on my laptop to see what I can do. I am not rich on clues so far.

**Comment by DaanHoogland** [comment]
I am getting rebase conflicts. In a meetup today but will find some time to resolve conflicts.

**Comment by DaanHoogland** [comment]
@rhtyd  @swill @NuxRo I resolved the conflict. Looks like we want to rerun integration tests and maybe test the dynamic service offering bit???

**Comment by NuxRo** [comment]
Good job, if integration tests pass I'll start testing it as well as the dynamic offering bit.

**Comment by DaanHoogland** [comment]
test_01_scale_vm failed again. Any body with the right infra to test this? @NuxRo ?

**Comment by NuxRo** [comment]
Not right now, I can try to build some nested stuff next week and have a go with that. I'll let you know.

**Comment by DaanHoogland** [comment]
@NuxRo, is for vmware as well or only xen?

**Comment by NuxRo** [comment]
Right, had a closer look. That check fails due to env problems:
https://paste.fedoraproject.org/367164/12782146/raw/

i.e. "updatevirtualmachine failed, due to: errorCode: 431, errorText:Unable to find network with trafficType Guest and guestType Shared in zone 1"

The check should be run in a Security Group zone, not in a pure Advanced one.

Talked to @swill and he doesn't know how to get out of this pickle.

@wido Do things still check out for you? I'm guessing the feature is mostly of interest to you (,me) and others who run public cloud with SG.

Any suggestions how to move forward?

**Comment by swill** [comment]
@DaanHoogland can you rebase and push this one again to see if we can get travis to finally pass?  thx...

**Comment by DaanHoogland** [comment]
@swill I just reopened. was there any reason to rebase?

**Comment by swill** [comment]
there were a couple fixes to jenkins and travis in a commit a few days ago, so i wanted to make sure those fixes were in the code base when you pushed.  I am concerned some of the PRs with failing Jenkins or Travis runs are missing some of the recent fixes.  I am going to have to make judgement calls on some of these anyway.  Just trying to get a final status on everything before I merge and freeze.

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 1
   Errors: 0
 Duration: 4h 17m 48s
```

**Summary of the problem(s):**
```
FAIL: test_03_vpc_privategw_restart_vpc_cleanup (integration.smoke.test_privategw_acl.TestPrivateGwACL)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs2/cloudstack/test/integration/smoke/test_privategw_acl.py", line 274, in test_03_vpc_privategw_restart_vpc_cleanup
    self.performVPCTests(vpc_off, restart_with_cleanup = True)
  File "/data/git/cs2/cloudstack/test/integration/smoke/test_privategw_acl.py", line 351, in performVPCTests
    self.check_pvt_gw_connectivity(vm1, public_ip_1, [vm2.nic[0].ipaddress, vm1.nic[0].ipaddress])
  File "/data/git/cs2/cloudstack/test/integration/smoke/test_privategw_acl.py", line 704, in check_pvt_gw_connectivity
    "Ping to VM on Network Tier N from VM in Network Tier A should be successful at least for 2 out of 3 VMs"
AssertionError: Ping to VM on Network Tier N from VM in Network Tier A should be successful at least for 2 out of 3 VMs
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_BW9J52/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_17_2016_17_19_42_5JBV7P:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_17_2016_17_19_42_5JBV7P/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_17_2016_17_19_42_5JBV7P/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_17_2016_17_19_42_5JBV7P/runinfo.txt)

**`/tmp/MarvinLogs/test_network_BW9J52:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_BW9J52/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_BW9J52/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_BW9J52/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_HFGHFG:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_HFGHFG/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_HFGHFG/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_HFGHFG/runinfo.txt)


Uploads will be available until `2016-07-17 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by NuxRo** [comment]
Meh, this PR really doesn't want to make the freeze.

Daan, any ideas?

**Comment by swill** [comment]
Running CI again on a new machine to see if the same thing pops since I have a free CI env right now.  We won't know till tomorrow morning (my time anyway).

**Comment by DaanHoogland** [comment]
@NuxRo the problem in @swill his CI run is new to me. The scale vm problem is xen specific. I'd have to debug it on xen based env. My bubble currently doesn't deploy xen based envs. I would have to spend some time on that.

**Comment by wido** [comment]
@swill Is the failure related to this PR you think? Having this in 4.9 would be a good to have.

**Comment by DaanHoogland** [comment]
@wido the travis problems seem to be in xen functionality. I haven't got a working xen setup atm. The eror from @swill his run is not specific to this PR

**Comment by wido** [comment]
@DaanHoogland Good! So that means there is no direct blocker for this PR, right?

This is a great addition for CloudStack which I personally would like to see go in.

**Comment by DaanHoogland** [comment]
I don't know who uses scalevm on xen @wido... @remibergsma, do you guys? anybosy else?

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 0
   Errors: 0
 Duration: 9h 11m 45s
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_18_2016_00_50_54_ZEE42A:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_18_2016_00_50_54_ZEE42A/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_18_2016_00_50_54_ZEE42A/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_18_2016_00_50_54_ZEE42A/runinfo.txt)

**`/tmp/MarvinLogs/test_network_76OYPC:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_76OYPC/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_76OYPC/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_76OYPC/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_9JCGWG:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_9JCGWG/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_9JCGWG/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_9JCGWG/runinfo.txt)


Uploads will be available until `2016-07-18 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
CI came back with a clean bill of health this time...

The scale-vm error was also seen in a different PR, so I don't think it is related to this one.  I think this one is ready and I just make a judgement call.  Our tooling is not 100% accurate and we will be doing physical hardware testing on the RC, so think this one gets in.  Thanks for the effort everyone...

**Comment by NuxRo** [comment]
Thanks buddies.

**Comment by swill** [comment]
This is one of the only PRs I have seen consistently failing on the scale vm test.  I am not sure what to think about that.  Can you rebase and force push or close and reopen to see if we can kick this off again to try again?  thx...

**Comment by DaanHoogland** [comment]
@swill I think it is genuine.

**Comment by swill** [comment]
@DaanHoogland you think the issue that travis is having is a genuine issue, is that what you mean?

**Comment by DaanHoogland** [comment]
Yes, I'm just not sure it is Travis or the xen plug-in.

**Comment by swill** [comment]
OK. I don't really have the bandwidth to troubleshoot this (unfortunately).  This is one of the many PRs that are on the bubble which I am trying to figure out what to do with. For now I am thinking not to merge it because this issue seems to be only happening with this PR. I am already almost a week late (of my own imposed deadline), so I have to make hard decisions soon on some of these.

**Comment by DaanHoogland** [comment]
I might find the time but it is long since I worked with xen. Also I must find new joy for this hobby. So no promises

**Comment by swill** [comment]
Have a nice weekend. :)

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -2451,6 +2455,21 @@ public UserVm updateVirtualMachine(long id, String displayName, String group, Bo
             isDynamicallyScalable = vm.isDynamicallyScalable();
         }
 
+        // Get default guest network in Basic zone
+        DataCenterVO zone = _dcDao.findById(vm.getDataCenterId());
+        Network defaultNetwork = _networkModel.getExclusiveGuestNetwork(zone.getId());
+
+        boolean isVmWare = (vm.getHypervisorType() == HypervisorType.VMware);
```
Comment (remibergsma) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: Small detail: Please check the spelling of VMware, I see lots of different variations and believe VMware is the right one.

**Comment by DaanHoogland** [comment]
@swill, i had a nice weekend, thanks. The test that is failing throws a different kind of exception if I run it on master in my bubble. I have just pushed a limited part of the change to see what it does and if the exception is still the same.

**Comment by DaanHoogland** [comment]
@swill, the test_scale_vm didn't error out because it wasn't run on this version! I have lost a tiny bit of confidence in our travis runs again. I will now start adding the rest of the code again.

**Comment by DaanHoogland** [comment]
got the test to run locally with
```
ssh cs0 /data/shared/helper_scripts/cloudstack/build_run_deploy_test.sh -m /data/git/cs1/cloudstack/setup/dev/advanced.cfg -S -s -t -f /data/shared/helper_scripts/cloudstack/run_scale_vm.sh
```
, so the rest is office-time-work

**Comment by NuxRo** [comment]
Thanks Daan,

Does that mean we're good to go?

**Comment by DaanHoogland** [comment]
@NuxRo it means debugging the problem has started. what is in here now is not much (you can review it)

**Comment by DaanHoogland** [comment]
@NuxRo can you test if it is still working for you?
@ustcweizhou can you see if your fixes are still met?
@swill can you reschedule integration tests?

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 82
  Skipped: 0
   Failed: 0
   Errors: 3
 Duration: 8h 35m 03s
```

**Summary of the problem(s):**
```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestRVPCSite2SiteVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 835, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_TJUSR5/results.txt
```

```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcRemoteAccessVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 293, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_TJUSR5/results.txt
```

```
ERROR: test suite for <class 'integration.smoke.test_vpc_vpn.TestVpcSite2SiteVpn'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 209, in run
    self.setUp()
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/usr/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/util.py", line 471, in try_run
    return func()
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vpc_vpn.py", line 472, in setUpClass
    cls.template.download(cls.apiclient)
  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 1350, in download
    elif 'Downloaded' in template.status:
TypeError: argument of type 'NoneType' is not iterable
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_TJUSR5/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_22_2016_06_11_44_7O49QH:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_22_2016_06_11_44_7O49QH/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_22_2016_06_11_44_7O49QH/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/DeployDataCenter__May_22_2016_06_11_44_7O49QH/runinfo.txt)

**`/tmp/MarvinLogs/test_network_TJUSR5:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_TJUSR5/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_TJUSR5/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_network_TJUSR5/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_179GPQ:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_179GPQ/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_179GPQ/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1297/tmp/MarvinLogs/test_vpc_routers_179GPQ/runinfo.txt)


Uploads will be available until `2016-07-24 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
This is coming back clean now.  Those failures are not related to this PR.

**Comment by DaanHoogland** [comment]
@swill, do you know what those errors are about? (always a dizzy feeling when this happens)

**Comment by swill** [comment]
Apparently `status` is missing from the `listTemplates` response in some cases.  I will investigate after we freeze and before we cut an RC to see if we can resolve it.

**Pull request merged**